packages/alfa-dom/src/get-tab-sequence.ts

```diff
@@ -0,0 +1,40 @@
+import { getTabIndex } from "./get-tab-index";
+import { Element, Node, NodeType } from "./types";
+
+export interface TabIndexedElement extends Element {
+  index: number;
+}
+
+/**
+ * @see https://www.w3.org/TR/html/editing.html#the-tabindex-attribute
+ */
+export function getTabSequence(
+  element: Element,
+  tabSequence: Array<TabIndexedElement> = []
+): Array<TabIndexedElement> {
+  (<Array<Node>>element.childNodes).forEach(element => {
+    if (element.nodeType === NodeType.Element) {
+      getTabSequence(<TabIndexedElement>element, tabSequence);
+    }
+  });
+
+  const index = getTabIndex(element);
+
+  if (index === null || index === -1) {
+    return tabSequence;
+  }
+
+  const weightedElement = <TabIndexedElement>element;
+
+  if (index > 0) {
+    weightedElement.index = index - 1;
+  }
+
+  if (index === 0) {
+    weightedElement.index = Number.MAX_SAFE_INTEGER;
+  }
+
+  tabSequence.push(weightedElement);
+
+  return tabSequence.sort((a, b) => a.index - b.index);
+}
```

Comment:
While this implementation is correct–kudos for that!–it is not optimal. The time complexity is currently in the order of `O(n (n log n))` due to sorting the entire tab sequence on each recursive invocation of `getTabSequence()`. This can be brought down to `O(n log n)` in the following way:

1. Initialise an empty array `result` that will contain the tab sequence.
2. Use `traverseNode()` to traverse the given node and its children. For every node:
    1. Compute its tab index using `getTabIndex()`.
    2. If the tab index is different from `null` and greater than or equal to `0`, add the node to `result` using binary search.
3. Return `result`.